src/hooks/connectWeb3/useActiveWeb3React.tsx

```diff
@@ -56,43 +59,56 @@ function getWallet(provider?: JsonRpcProvider | Eip1193Provider) {
     } else if (JsonRpcProvider.isProvider((provider as any).provider)) {
       throw new Error('Eip1193Bridge is experimental: pass your ethers Provider directly')
     } else {
-      ;[connector, hooks] = initializeConnector((actions) => new EIP1193(actions, provider))
+      ;[connector, hooks] = initializeConnector((actions) => new EIP1193({ actions, provider }))
     }
     connector.activate()
     return { connector, hooks }
   }
   return EMPTY_STATE
 }
 
+export let connections: [Connector, Web3ReactHooks][] = []
+
 interface ActiveWeb3ProviderProps {
   jsonRpcEndpoint?: string | JsonRpcProvider
   provider?: Eip1193Provider | JsonRpcProvider
 }
 
 export function ActiveWeb3Provider({
   jsonRpcEndpoint,
-  provider,
+  provider: propsProvider,
   children,
 }: PropsWithChildren<ActiveWeb3ProviderProps>) {
+  const metaMaskConnection = useMemo(
+    () => toWeb3Connection(initializeConnector<MetaMask>((actions) => new MetaMask({ actions }))),
+    []
+  )
+  const walletConnectConnectionQR = useMemo(() => getWalletConnectConnection(false, jsonRpcEndpoint), [jsonRpcEndpoint]) // WC via tile QR code scan
+  const walletConnectConnectionPopup = useMemo(
+    () => getWalletConnectConnection(true, jsonRpcEndpoint),
+    [jsonRpcEndpoint]
+  ) // WC via built-in popup
+  connections = [metaMaskConnection, walletConnectConnectionQR, walletConnectConnectionPopup]
+
   const network = useMemo(() => getNetwork(jsonRpcEndpoint), [jsonRpcEndpoint])
-  const wallet = useMemo(() => getWallet(provider), [provider])
+  const activeProvider = useActiveWalletProvider()
+  const wallet = useMemo(() => getWallet(propsProvider ?? activeProvider), [propsProvider, activeProvider])
 
   // eslint-disable-next-line prefer-const
-  let { connector, hooks } = wallet.hooks.useIsActive() || network === EMPTY_STATE ? wallet : network
+  let { connector, hooks } = wallet !== EMPTY_STATE ? wallet : network
   let accounts = hooks.useAccounts()
   let account = hooks.useAccount()
   let activating = hooks.useIsActivating()
   let active = hooks.useIsActive()
   let chainId = hooks.useChainId()
-  let error = hooks.useError()
   let library = hooks.useProvider()
 
   const web3 = useMemo(() => {
     if (connector === EMPTY || !(active || activating)) {
       return EMPTY_WEB3
     }
-    return { connector, library, chainId, accounts, account, active, activating, error }
-  }, [account, accounts, activating, active, chainId, connector, error, library])
+    return { connector, library, chainId, accounts, account, active, activating }
+  }, [account, accounts, activating, active, chainId, connector, library])
 
   const updateWeb3 = (updateContext: Web3ContextType) => {
```

Comment:
I'm afraid this might cause frequent updates. Can the updater be moved to a separate context?

Reply:
hm - could you clarify this? wdym by "risk extra renders by instantiating objects multiple times"

Reply:
This is being recreated on each render, and then the value of the provider is being instantiated on each render.

Instead, make updateWeb3 a callback
```
const updateWeb3 = useCallback((updateContext: Web3ContextType) => {
```

and memoize the value so it is not recreated every time the provider renders
```
const value = useMemo(() => ({ web3, updateWeb3 }), [web3, updateWeb3])
return <Web3Context.Provider value={value}>{children}</Web3Context.Provider>
```